### DISCUSSION TITLE
It doesn't work with "ignore" option in CopyWebpackPlugin

### DISCUSSION THREAD

**Discussion by 17621802479** [webpack/webpack#14216] [Q&A]
hello everyone,here I meet an bug of "ignore" option in CopyWebpackPlugin;
flowing is my code of webpack.config.js,but when I build my code, I get a error like "ERROR in Conflict: Multiple assets emit different content to the same filename index.html".I guess it is because "ignore" option in CopyWebpackPlugin does not work,so I replace the code "ignore: path.join(__dirname, "public/index.html"), by "ignore: **/index.html," ,then it works. my env is windows, it is strange that  the code "ignore: path.join(__dirname, "public/index.html") has no problem in linux env, so ,who can tell me why, why the bug exist,and why the code of "ignore: **/index.html," can work,and why  linux env can work well? thank you ~~

`/* eslint-disable global-require */
// __dirname是node.js中的一个全局变量，它指向当前执行脚本所在的目录
// path是node.js中提供的处理文件路径的小工具。 (http://www.runoob.com/nodejs/nodejs-path-module.html)
const path = require("path");
const fs = require("fs");
const os = require("os");
const packageinfo = require("./package.json");

const appDirectory = fs.realpathSync(process.cwd());
const resolveApp = (relativePath) => path.resolve(appDirectory, relativePath);

const chokidar = require("chokidar");
const webpack = require("webpack");

const errorOverlayMiddleware = require("react-dev-utils/errorOverlayMiddleware");
const evalSourceMapMiddleware = require("react-dev-utils/evalSourceMapMiddleware");
// const noopServiceWorkerMiddleware = require("react-dev-utils/noopServiceWorkerMiddleware");

const proxySetup = resolveApp("src/setupProxy.js");

const HtmlWebpackPlugin = require("html-webpack-plugin"); // 引入html-webpack-plugin插件
// webpack.config.js  这里贴出来的是和 `css` 相关的配置，不是完整配置，需要源码可以在文章末尾留言
const MiniCssExtractPlugin = require("mini-css-extract-plugin"); // CSS 抽离插件
const TerserPlugin = require("terser-webpack-plugin"); // 压缩js包
const CssMinimizerPlugin = require("css-minimizer-webpack-plugin");
const HappyPack = require("happypack");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const { BundleAnalyzerPlugin } = require("webpack-bundle-analyzer");
// 基础库不打包，直接CDN引入
// const HtmlWebpackExternalsPlugin = require("html-webpack-externals-plugin");
const CopyWebpackPlugin = require("copy-webpack-plugin");

const htmlTempPath = "public/index.html";
const outPath = "build/";

module.exports = (env) => {
  console.log(os.platform());
  const osName = os.platform();
  let mode = "development";
  const isProd = env && !!env.production;
  // console.log( env?.production);
  if (isProd) {
    mode = "production";
  }

  const prodPlugins = [
    osName === "darwin" && new BundleAnalyzerPlugin(),
    new CleanWebpackPlugin(),
  ].filter(Boolean);
  const otherPlugins = isProd ? prodPlugins : [];

  let splitChunks = {};
  if (isProd) {
    splitChunks = {
      minSize: 0, // 形成一个新代码块最小的体积
      automaticNameDelimiter: "~", // 打包分割符
      cacheGroups: {
        ffff: {
          // 基本框架
          chunks: "all",
          test: /(react|react-dom|react-dom-router|react-redux|redux-saga|redux-thunk)/,
          priority: 90,
          name: "ffff",
        },
        material: {
          // 将体积较大的d3单独提取包，指定页面需要的时候再异步加载
          chunks: "all",
          minChunks: 2,
          test: /[\\/]node_modules[\\/]_?@material-ui(.*)/,
          priority: 100, //
          name: "material",
        },
        echarts: {
          // 将体积较大的d3单独提取包，指定页面需要的时候再异步加载
          chunks: "all",
          test: /[\\/]node_modules[\\/]_?echarts(.*)/,
          priority: 110, //
          name: "echarts",
        },
        draftjs: {
          // 将体积较大的d3单独提取包，指定页面需要的时候再异步加载
          chunks: "all",
          test: /[\\/]node_modules[\\/]_?draft-js(.*)/,
          priority: 120, //
          name: "draftjs",
        },
        commons: {
          // 其余同步加载包
          chunks: "all",
          minChunks: 2,
          name: "commons",
          priority: 80,
        },
      },
    };
  }

  return {
    mode,
    // 项目入口，webpack从此处开始构建
    entry: {
      // "webpack/hot/poll?1000",
      main: [path.join(__dirname, "src/index.js")], // 指定入口，可以指定多个。参考webpack文档
    },
    output: {
      path: path.join(__dirname, outPath), // bundle生成(emit)到哪里
      publicPath: "/",
      filename: "[name].js", // bundle生成文件的名称
    },
    performance: {
      hints: "warning", // 枚举
      maxAssetSize: 30000000, // 整数类型（以字节为单位）
      maxEntrypointSize: 50000000, // 整数类型（以字节为单位）
    },
    module: {
      rules: [
        {
          test: /\.m?js$/,
          resolve: {
            fullySpecified: false, // disable the behaviour
          },
        },
        // {
        //   test: /\.(tsx|ts)$/,
        //   include: [path.resolve(__dirname, "docs")],
        //   loader: path.resolve(__dirname, "./webpack/codeLoader.js"), // "raw-loader"
        // },
        {
          test: /\.(tsx|ts)$/,
          include: [path.resolve(__dirname, "src"), path.resolve(__dirname, "lib")],
          use: ["ts-loader"], // "raw-loader"
        },

        // {
        //   test: /\.(tsx|ts)$/,
        //   loader: "awesome-typescript-loader",
        //   include: [path.resolve(__dirname, "src"), path.resolve(__dirname, "lib")],
        //   exclude: [/node_modules/, path.resolve(__dirname, outPath)],
        // },
        {
          test: /\.(jsx|js)$/,
          loader: "happypack/loader",
          include: [path.resolve(__dirname, "src"), path.resolve(__dirname, "lib")],
          options: {
            cacheDirectory: true,
          },
        },

        {
          test: /\.css$/,
          include: [
            path.resolve(__dirname, "src"),
            path.resolve(__dirname, "src/pages/Demo/"),
            path.resolve(__dirname, "./node_modules/react-push-notification/dist/notifications/"),
            path.resolve(__dirname, "./node_modules/react-push-notification/dist/notifications/"),
            path.resolve(__dirname, "./node_modules/codemirror/lib/"),
            path.resolve(__dirname, "./node_modules/codemirror/theme/"),
            path.resolve(__dirname, "./node_modules/codemirror/addon/"),

            // path.resolve(__dirname, "./node_modules/draft-js/dist"),
          ], // test、include、exclude 缩小查找范围
          use: [
            {
              loader: MiniCssExtractPlugin.loader,
            }, // 放在css-loader前
            {
              loader: "css-loader",
              options: {
                url: false,
                sourceMap: !isProd,
              },
            }, // 影响所有依赖css-loader的插件生成sourceMap
          ],
        },
        {
          test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
          loader: "url-loader",
          options: {
            esModule: false,
            limit: 10000, // 小于10KB图片，转base64编码
            name: "img/[name].[hash:7].[ext]",
          },
        },
        {
          test: /\.png$/i, // 以png为例
          use: [
            "file-loader",
            {
              loader: "image-webpack-loader",
              options: {
                pngquant: {
                  quality: [0.65, 0.9], // 设置png的品质区间
                },
              },
            },
          ],
        },
        {
          test: /\.(svg|jpg|gif|jpeg)$/,
          loader: "file-loader",
        },
        {
          test: /\.(woff|woff2|eot|ttf|otf)$/,
          loader: "url-loader",
        },
      ],
    },

    plugins: [
      new webpack.DefinePlugin({ "process.env": `{v:'${packageinfo.version}'}` }),
      new webpack.BannerPlugin(`pacakge version：${packageinfo.version}`),
      // 在plugins使用插件
      new HtmlWebpackPlugin({
        filename: "index.html", // 生成的文件名
        template: htmlTempPath, // 被打包的源文件路径
        hash: true,
        chunks: ["material", "echarts", "draftjs", "commons", "ffff", "main"],
      }),
      new HappyPack({
        loaders: [
          {
            loader: "babel-loader",
            options: {
              cacheDirectory: true,
            },
          },
        ],
        threads: 4,
      }),
      new MiniCssExtractPlugin({
        filename: "style.[contenthash].css", // 指定文件名 生产环境可写成 style.[contenthash].css 避免缓存问题
      }),
      new webpack.HotModuleReplacementPlugin(),
      // new HtmlWebpackExternalsPlugin({
      //   externals: [
      //     {
      //       module: "react",
      //       entry: "http://xxx/react.production.min.js",
      //       global: "React",
      //     },
      //     {
      //       module: "react-dom",
      //       entry: "http://xxx/react-dom.production.min.js",
      //       global: "ReactDOM",
      //     },
      //   ],
      // }),
      new CopyWebpackPlugin({
        patterns: [
          {
            from: path.join(__dirname, "public"),
            to: path.join(__dirname, outPath),
            toType: "dir",
            globOptions: {
              ignore: path.join(__dirname, "public/index.html"),
            },
          },
        ],
      }),
      ...otherPlugins,
    ],
    // 三方库分离
    optimization: {
      minimize: isProd, // 可省略，默认最优配置：生产环境，压缩 true。开发环境，不压缩 false
      minimizer: [
        new TerserPlugin({
          parallel: true, // 可省略，默认开启并行
          terserOptions: {
            toplevel: true, // 最高级别，删除无用代码
            ie8: false,
            safari10: false,
          },
        }),
        // new CssMinimizerPlugin({
        //   parallel: true, // 可省略，默认开启并行
        //   minimizerOptions: {
        //     preset: "advanced", // 需额外安装
        //   },
        // }),
      ],
      // 公共脚本分离
      splitChunks,
    },
    resolve: {
      extensions: [".js", ".jsx", ".ts", ".tsx", "mdx", ".json", ".css"],
      fallback: {
        fs: false,
        tls: false,
        net: false,
        zlib: false,
        http: false,
        https: false,
        assert: false,
        "crypto-browserify": require.resolve("crypto-browserify"), // if you want to use this module also don't forget npm i crypto-browserify
        crypto: require.resolve("crypto-browserify"),
        path: require.resolve("path-browserify"),
        url: require.resolve("url"),
        buffer: false,
        util: false,
        stream: require.resolve("stream-browserify"),
      },
      alias: {
        "@/lib": path.resolve(__dirname, "src/lib"),
        "@/pages": path.resolve(__dirname, "src/pages"),
        "@/utils": path.resolve(__dirname, "src/utils"),
        "@/hooks": path.resolve(__dirname, "src/hooks"),
        "@/components": path.resolve(__dirname, "src/components"),
        "@/api": path.resolve(__dirname, "src/api"),
        "@/constants": path.resolve(__dirname, "src/constants"),
        "@/redux": path.resolve(__dirname, "src/redux"),
      },
    },

    devtool: isProd ? "source-map" : "source-map",
    // devtool: "inline-source-map",
    devServer: {
      port: 3050,
      host: "localhost",
      inline: true, // 实时刷新
      hot: true, // 模块热替换机制
      quiet: false,
      contentBase: path.resolve(__dirname, outPath), // 此处的路径必须和输出output文件的路径一致 否则无法自动更新，或者是基于output的相对路径
      compress: true, // gzip压缩
      watchContentBase: true,
      watchOptions: {
        aggregateTimeout: 200,
        poll: 1000,
        ignored: [".history", ".vscode", ".husky", "coverage", "build", "**/node_modules"],
      },
      historyApiFallback: {
        // Paths with dots should still use the history fallback.
        // See https://github.com/facebook/create-react-app/issues/387.
        disableDotRule: true,
      },
      progress: false,
      publicPath: "/",
      open: true, // 当open启用时，开发服务器将打开浏览器。
      overlay: {
        // 当存在编译器错误或警告时，在浏览器中显示全屏覆盖,显示警告和错误：
        warnings: true,
        errors: true,
      },
      before(app, server) {
        const files = [
          path.resolve(htmlTempPath), // 与 HtmlWebpackPlugin 的 template 对应
        ];
        chokidar.watch(files).on("all", () => {
          server.sockWrite(server.sockets, "content-changed");
        });
        if (fs.existsSync(proxySetup)) {
          // This registers user provided middleware for proxy reasons
          require(proxySetup)(app);
        }
        // This lets us fetch source contents from webpack for the error overlay
        app.use(evalSourceMapMiddleware(server));
        // This lets us open files from the runtime error overlay.
        app.use(errorOverlayMiddleware());

        // This service worker file is effectively a 'no-op' that will reset any
        // previous service worker registered for the same host:port combination.
        // We do this in development to avoid hitting the production cache if
        // it used the same host and port.
        // https://github.com/facebook/create-react-app/issues/2272#issuecomment-302832432
      },
    },
    cache: {
      type: "filesystem",
      // 可选配置
      buildDependencies: {
        config: [__filename], // 当构建依赖的config文件（通过 require 依赖）内容发生变化时，缓存失效
      },
      name: "development-cache",
    },
  };
};
`

**Comment by alexander-akait** [comment]
Sorry the issue is not readable, please use github repo to provide example of the problem
